### PR TITLE
fix: Closure context

### PR DESCRIPTION
**Pull request by scolsen**
This PR does two things:

1. Adds `Semigroup` instances for `Env` and `Context` to offer a way to combine each of these types, respectively. Envs are combined using the left-biased union of the maps containing their bindings.
2. Updates our handling of closures to evaluate them under the *semigroup combination* of the context captured during their creation and the context under which they are evaluated, biasing toward the creation-context. This allows us to ensure closures have access to the latest global definitions, fixing #1101, while also ensuring that, in the case of conflicts, they prefer the local definitions captured during the creation context.

### CODE REVIEW

**Comment by scolsen** [comment]
@hellerve hopefully this unblocks you for `derive`! I used the example you included in the issue and it works after this change.

**Comment by scolsen** [comment]
This fixes #1101 as is, but thinking about this a bit more, the semigroup for `Env` likely has to be a bit more nuanced.

We probably want to left-bias for *internal Envs* but right bias for other envs--that is, any `let` bindings in a closure context should win over global bindings of the same name, but any global bindings in the closure context should lose against global bindings of the same name in the latter evaluation/application context.

In other words, the semigroup for `Env` should right-bias global envs and type envs and left-bias internal envs.

**Comment by eriksvedang** [comment]
> This fixes #1101 as is, but thinking about this a bit more, the semigroup for `Env` likely has to be a bit more nuanced.
> 
> We probably want to left-bias for _internal Envs_ but right bias for other envs--that is, any `let` bindings in a closure context should win over global bindings of the same name, but any global bindings in the closure context should lose against global bindings of the same name in the latter evaluation/application context.
> 
> In other words, the semigroup for `Env` should right-bias global envs and type envs and left-bias internal envs.

I haven't thought hard about this but given that you're correct it seems worth fixing before merge:ing, right?

File: src/Eval.hs [hunk 1]
```diff
@@ -302,7 +302,7 @@ eval ctx xobj@(XObj o info ty) preference resolver =
                 (newCtx, evaledArgs) <- foldlM successiveEval (ctx, Right []) args
                 case evaledArgs of
                   Right okArgs -> do
-                    (_, res) <- apply c{contextHistory=contextHistory ctx} body params okArgs
+                    (_, res) <- apply (c {contextHistory = contextHistory ctx} <> ctx) body params okArgs
```
Comment (eriksvedang) [inline src/Eval.hs:305]: So clean, I'm impressed!

**Comment by scolsen** [comment]
> > This fixes #1101 as is, but thinking about this a bit more, the semigroup for `Env` likely has to be a bit more nuanced.
> > We probably want to left-bias for _internal Envs_ but right bias for other envs--that is, any `let` bindings in a closure context should win over global bindings of the same name, but any global bindings in the closure context should lose against global bindings of the same name in the latter evaluation/application context.
> > In other words, the semigroup for `Env` should right-bias global envs and type envs and left-bias internal envs.
> 
> I haven't thought hard about this but given that you're correct it seems worth fixing before merge:ing, right?

Yes, if I am right I think we should wait until this is updated! Let me run a couple checks to see if my thinking is right here, and I'll let you know

**Comment by scolsen** [comment]
My assumption was right--made the fix and added some tests to check some basic closure behaviors:

```
./test/closure.carp
Test 'closures over global variables get the global variable's latest state.' passed
Test 'closures with arguments use argument values' passed
Test 'closures with let bindings prefer let bindings over global names' passed
Test 'dynamic closures can refer to global bindings delcared after the closure' passed
Test 'nested closures prefer closed-over bindings' passed
Results:
	|=====|
	Passed: 5	Failed: 0
```

**Comment by scolsen** [comment]
@eriksvedang should be ready now assuming all tests pass (passed locally)

File: src/Eval.hs (hunk 1 above)
Comment (scolsen) [reply src/Eval.hs:305]: Thanks!

**Comment by eriksvedang** [comment]
Really nice! I'm a little confused though, the tests seem to be mainly for statically compiled closures, no? (which is not what the patch addresses)

**Comment by scolsen** [comment]
> Really nice! I'm a little confused though, the tests seem to be mainly for statically compiled closures, no? (which is not what the patch addresses)

Doh! Thanks for catching that. Looks like I got confused this morning. Added some commits to change it to dynamics
